Feature/project skeleton

andrey-dihtiarov opened the pull request.

andrey-dihtiarov committed:
project-skeleton(feat): new dependencies added
- node-sass added
- history added

andrey-dihtiarov committed:
project-skeleton(feat): css files changed to scss

andrey-dihtiarov committed:
project-skeleton(feat): new forms added
- AccountForm
- CapabilitiesForm
- ContactsForm
- ProfileForm

andrey-dihtiarov committed:
project-skeleton(feat): new constants added
- routes constants
- general general constants

andrey-dihtiarov committed:
project-skeleton(feat): new pages added
- HomePage
- NotFoundPage
- EditUserPage
- NewUserPage
- UserPage
- UsersPage

andrey-dihtiarov committed:
project-skeleton(feat): Header component added

andrey-dihtiarov committed:
project-skeleton(feat): MainLayout component added

andrey-dihtiarov committed:
project-skeleton(feat): createBrowserHistory added

andrey-dihtiarov committed:
project-skeleton(feat): all required routes added
- HomeRoute
- UserRoute
- UsersRoute
- NewUserRoute
- EditUserRoute
- NotFoundRoute

andrey-dihtiarov committed:
project-skeleton(feat): Routes component added to App component

andrey-dihtiarov committed:
project-skeleton(feat): changed styles import file to scss

oleksandrbazil commented on package.json:
```diff
@@ -8,7 +8,8 @@
     "formik": "^2.2.6",
-    "node-sass": "^6.0.0",
+    "history": "^5.0.0",
+    "node-sass": "4.14.1",
```
а зачем в проекте эти изменения?

andrey-dihtiarov replied:
Моя текущая нода(14.17.0) не поддерживает node-sass 5 и 6. И из-за этого поставил 4ю версию.
![image](https://user-images.githubusercontent.com/82100556/119095006-67e66300-ba1a-11eb-8210-3c74c46f8c35.png)

andrey-dihtiarov replied:
Моя версия ноды(14.17.0) не поддерживает `node-sass` 5й и 6й версий.
![image](https://user-images.githubusercontent.com/82100556/119097117-0a074a80-ba1d-11eb-81ea-7ca579da33a3.png)

oleksandrbazil replied:
впервые вижу что бы занижали версии пакетов потому что локальная среда не тянет новые
Для тестового задания так может и можно делать, но в реальной практике занижать можно только когда есть несовместимость с другими зависимостями

oleksandrbazil replied:
насколько реально поднять локальное окружение?
и я не знаю зачем нам sass если мы будем использовать react modules

oleksandrbazil commented on src/constants/routes.js:
```diff
@@ -0,0 +1,5 @@
+export const ROUTE_HOME = '/';
```
а зачем выносить роуты в константы?

andrey-dihtiarov replied:
Подумал, что так будет удобнее. Особенно когда нужно будет в проекте делать пуш на какой-то роут. Плюс меньше опечаток, так как все определено как константа.

oleksandrbazil replied:
мысль кажется хорошей, но только на первый взгляд и вот почему:
- отдельный файл для констант  - ОК, но зачем тогда нам префикс ROUTE_?  - не вижу смысла, лучше что-то одно
- слабо верю что можно пушить пользователя на константу`/users/:id`
- можно не делать отдельный файл и импортировать всегда 1 константу, попробуй найти решение

oleksandrbazil commented on src/index.js:
```diff
@@ -1,6 +1,6 @@
 import React from 'react';
 import ReactDOM from 'react-dom';
-import './index.css';
+import './index.scss';
```
группируй импорты https://eslint.org/docs/rules/sort-imports
если я не ошибаюсь импорт стилей будет после всех остальных импортов, но точно не между компонентами

andrey-dihtiarov replied:
Да, забыл поправить. Вообще так и стараюсь делать. Но линтер на это не ругнется кстати. Он ругнется если я, допустим, `import React from 'react'` вниз поставлю.
![image](https://user-images.githubusercontent.com/82100556/119096313-1b038c00-ba1c-11eb-8ee6-72ab3d0cea96.png)

oleksandrbazil replied:
как правило реакт зависимость импортируют в самом вверху,
потом компоненты, 
потом модули
потом стили

oleksandrbazil commented on src/pages/users/NewUserPage/index.jsx:
```diff
@@ -0,0 +1,21 @@
+
+import { FORMS } from '../../../constants/general';
+
+const NewUserPage = () => {
```
тебе не кажется что этот компонент очень похож на src/pages/users/EditUserPage/index.jsx ?
Соответсвует ли это принципу DRY?

andrey-dihtiarov replied:
Да, согласен. Поправлю.

oleksandrbazil commented on src/routes/history.js:
```diff
@@ -0,0 +1,3 @@
+import { createBrowserHistory } from 'history';
+
+export default createBrowserHistory();
```
а зачем нам история браузера?

andrey-dihtiarov replied:
Да, наверное стоит это убрать и просто использовать `BrowserRouter`. А для истории ж есть `useHistory` хук.

oleksandrbazil reviewed (commented):
запроси ревью когда запушиш правки в репозиторий

andrey-dihtiarov committed:
project-skeleton: eslint rules updated

andrey-dihtiarov committed:
project-skeleton: imports order changed

andrey-dihtiarov committed:
project-skeleton: unnecessary "history" dependency removed

andrey-dihtiarov committed:
project-skeleton: routes file deleted

andrey-dihtiarov committed:
project-skeleton: general.js renamed to index.js, routes constants added

andrey-dihtiarov committed:
project-skeleton: removed EditUserPage and newUserPage and added UserFormPage

andrey-dihtiarov committed:
project-skeleton: Routes
- Router changed to BrowserRouter
- NewUser Page and EditUserPage changed to UserFormPage
- routes constants updated

oleksandrbazil commented on src/constants/index.js:
```diff
@@ -0,0 +1,35 @@
+import AccountForm from '../components/forms/AccountForm';
+import ProfileForm from '../components/forms/ProfileForm';
+import ContactsForm from '../components/forms/ContactsForm';
+import CapabilitiesForm from '../components/forms/CapabilitiesForm';
```
в компопнет UserFormPage >>>

oleksandrbazil commented on src/pages/users/UserFormPage/index.jsx:
```diff
@@ -0,0 +1,30 @@
+  return (
+    <div>
+      {isEditing ? 'Edit User Page' : 'New User Page'}
+      <div>
```
лишний див

oleksandrbazil commented on src/constants/index.js:
```diff
@@ -0,0 +1,35 @@
+export const FORMS = [
+  {
+    component: AccountForm,
+    route: 'account',
```
slug

oleksandrbazil commented on src/constants/index.js:
```diff
@@ -0,0 +1,35 @@
+  {
+    component: AccountForm,
+    route: 'account',
+    form: 'Account',
```
title

oleksandrbazil commented on src/pages/users/UserFormPage/index.jsx:
```diff
@@ -0,0 +1,30 @@
+      <div>
+        <Switch>
+          {FORMS.map(({ component: Form, route, form }) => (
+            <Route key={form} exact path={`${path}/${route}`} component={Form} />
```
slug

andrey-dihtiarov committed:
project-skeleton: FORMS constant moved to UserFormPage, changed form to slug

andrey-dihtiarov committed:
project-skeleton: unnecessary div removed
